### ISSUE TITLE
Issue calling MMAMS from the interface

### ISSUE THREAD

**Issue by NicholasRui** [amusecode/amuse#734]
Hi all,

I have been trying to use the AMUSE implementation of the merger code MMAMS ([Gaburov 2008](https://onlinelibrary.wiley.com/doi/full/10.1111/j.1745-3933.2007.00399.x)) in order to combine two stellar profiles which were evolved externally using MESA.

Due to difficulties installing all of AMUSE, I had installed amuse-framework together with amuse-mmams to use this package by itself. I created two particles and added the layers of both stars using the "add_shell" method, but I am getting the error below. I'm not sure whether this is an issue with the installation or with my input, but this occurs for a number of different inputs, and the issue is not the order in which I add the shells.

Thanks in advance.

```
---------------------------------------------------------------------------
CodeException                             Traceback (most recent call last)
/home/nrui/.local/lib/python3.6/site-packages/amuse/rfi/core.py in __call__(self, *arguments_list, **keyword_arguments)
    119
--> 120             dtype_to_result = self.interface.channel.recv_message(call_id, self.specification.id, handle_as_array)
    121         except Exception as ex:

/home/nrui/.local/lib/python3.6/site-packages/amuse/rfi/channel.py in recv_message(self, call_id, function_id, handle_as_array, has_units)
   1946
-> 1947         message.receive(self.socket)
   1948

/home/nrui/.local/lib/python3.6/site-packages/amuse/rfi/channel.py in receive(self, socket)
   1497
-> 1498         header_bytes = self._receive_all(44, socket)
   1499

/home/nrui/.local/lib/python3.6/site-packages/amuse/rfi/channel.py in _receive_all(self, nbytes, thesocket)
   1482             if len(data_bytes) == 0:
-> 1483                 raise exceptions.CodeException("lost connection to code")
   1484

CodeException: lost connection to code

During handling of the above exception, another exception occurred:

CodeException                             Traceback (most recent call last)
<ipython-input-2-36f82fcc752b> in <module>()
----> 1 make_mesa_dir_merger_from_mmams(1.5, 1.0, 'TAMSR125')

/home/nrui/merger_project/mergers.py in make_mesa_dir_merger_from_mmams(initial_mass, accretion_mass, accretion_loc, scale_loc)
   3601     pdb.set_trace()
   3602
-> 3603     interface.merge_two_stars(index_of_the_product=2, index_of_the_primary=0, index_of_the_secondary=1)
   3604     pdb.set_trace()
   3605

/home/nrui/.local/lib/python3.6/site-packages/amuse/rfi/core.py in __call__(self, *arguments_list, **keyword_arguments)
    121         except Exception as ex:
    122             CODE_LOG.info("Exception when calling function '{0}', of code '{1}', exception was '{2}'".format(self.specification.name, type(self.interface).__name__, ex))
--> 123             raise exceptions.CodeException("Exception when calling function '{0}', of code '{1}', exception was '{2}'".format(self.specification.name, type(self.interface).__name__, ex))
    124
    125         result = self.converted_results(dtype_to_result, handle_as_array)

CodeException: Exception when calling function 'merge_two_stars', of code 'MMAMSInterface', exception was 'lost connection to code'
```

**Comment by rieder** [comment]
Could you check in `/opt/etc/macports/macports.conf` if "universal_archs" is uncommented and set to "x86_64"?

**Comment by rieder** [comment]
Could you start the code with `redirection="none"` in addition to the other arguments? E.g.
```python
code = MMAMS(..., redirection="none")
```

**Comment by rieder** [comment]
The exception above occurs when a code crashes, with redirection="none" it may be clearer what causes this to happen.

**Comment by NicholasRui** [comment]
Thanks, I've done this. The code terminates almost immediately, and the output of the code is as follows:
![image](https://user-images.githubusercontent.com/25237984/109083956-b86c4f80-76bb-11eb-8a2a-f079fbb20513.png)

where I have added stellar shells assuming quantities in cgs. If I instead input these quantities in solar units, the code runs indefinitely and prints the same line over and over with a single item iterated.
![image](https://user-images.githubusercontent.com/25237984/109084067-e9e51b00-76bb-11eb-847f-098744b9306d.png)

After assigning 
```
    interface = mmams.MMAMSInterface(redirection='none')

    index_of_the_particle = [0, 1]
    profile_list = [profile1, profile2]
    star_mass = [initial_mass, accretion_mass]

    for ii in range(len(profile_list)):
        profile = profile_list[ii]

        interface.new_particle(mass=star_mass[ii])

        # Extract required quantities, arrange it into greatest r to least and then
        # add the shells in reverse order afterwards.
        radius = rsun * profile['radius']
        radius0 = np.append(radius, [0])
        density = 10 ** profile['logRho']
        volume_shell = (4 / 3) * np.pi * (radius0[:-1] ** 3 - radius0[1:] ** 3)
        d_mass = np.array(density * volume_shell)
        cumul_mass = np.flip(np.cumsum(np.flip(d_mass, axis=0)), axis=0)

        pressure = 10 ** profile['logP']
        temperature = 10 ** profile['logT']
        luminosity = lsun * profile['luminosity']
        molecular_weight = profile['mu']

        H1 = profile['h1']
        He4 = profile['he4']
        C12 = profile['c12']
        O16 = profile['o16']

        for jj in range(len(profile)):
            index = -jj-1
            interface.add_shell(index_of_the_particle = index_of_the_particle[ii],
                                d_mass = d_mass[index],
                                cumul_mass = cumul_mass[index],
                                radius = radius[index],
                                density = density[index],
                                pressure = pressure[index],
                                temperature = temperature[index],
                                luminosity = luminosity[index],
                                molecular_weight = molecular_weight[index],
                                H1 = H1[index],
                                He4 = He4[index],
                                C12 = C12[index],
                                N14 = 0,
                                O16 = O16[index],
                                Ne20 = 0,
                                Mg24 = 0,
                                Si28 = 0,
                                Fe56 = 0
                                )

    interface.merge_two_stars(index_of_the_product=2, index_of_the_primary=0, index_of_the_secondary=1)
```

Am I calling this function incorrectly? Alternatively, is there something wrong with my input (it's a fairly standard and non-pathological MESA run of not very massive stars)? Is there any documentation which describes how to correctly call this function without running MESA directly in AMUSE?

**Comment by NicholasRui** [comment]
Upon closer examination, it looks like the pressure bracketing is failing because of the following lines in amuse/src/amuse/community/mmams/src/mmas2/src/mmas/merge_stars_gsl_adaptive.cpp:

```
if (params.negative_pressure == 1)  {
      pressure.push_back(-p_unit);
      return NEGATIVE_PRESSURE;
}
```

This flag always gets triggered and appends -p_unit = -GM^2/R^4 to the end of the pressure vector, preventing the pressure root bracketing. If I attempt to comment this part of the code out (so that this "ghost" pressure entry is not appended to the pressure vector), I get a separate issue where the central pressure attempts to converge to a very small value.

I'm not sure what the issue is. I have specified all my units in cgs; the NEGATIVE_PRESSURE flag being triggered appears to suggest something wrong with the input, but I'm not totally sure what it is, or what the NEGATIVE_PRESSURE flag is signifying.

Attached below are the two stellar profile files which are passed shell-by-shell into the MMAMS interface in AMUSE.
[profile1.txt](https://github.com/amusecode/amuse/files/6061456/profile1.txt)
[profile2.txt](https://github.com/amusecode/amuse/files/6061455/profile2.txt)

As an additional note, I have figured out that AMUSE uses the file "merge_stars_gsl_adaptive.cpp." However, it seems to contain some other scripts, "merge_stars.cpp," "merge_stars_gsl.cpp," "merge_stars_consistently.cpp," which are all different from each other. As far as I can tell, I think they're unused, so it may also be an idea to try to use these, but I'm not sure what the difference is, or why the default would not have been one of these others.
